More exposition on motivation and use cases. (#13)
fl1ger/deleg: draft-dnsop-deleg-transport.md

"OPTIONAL" in this document are to be interpreted as described in
BCP 14 {{?RFC2119}} {{?RFC8174}} when, and only when, they appear in
all capitals, as shown here.

Terminology regarding the Domain Name System comes from {{?BCP219}}, with addition terms defined here:

* transport parameters	  The set of attributes that define how to reach the nameserver specified in a ServiceMode DELEG record.
* [tbd]

## Introductory Examples

To use a DELEG record to point a resolver to an authority that supports DNS-over-TLS {{?RFC7858}}, this example shows a possible response from an authoritative server in the Authority section of the DNS response:

    example.com. 86400  IN DELEG 1 ns2.example.com. (
                    alpn=dot
                    tlsa="2 0 1 ABC..."
                    ipv4hint=192.0.2.54
                    ipv6hint=2001:db8:2423::3 )
    example.com. 86400  IN NS ns1.example.com.

The alpn attribute specifies DNS-over-TLS, with the tlsa attribute
providing the necessary cryptographic material to establish the
connection.  The ipv4hint and ipv6hint serves the traditional role of
glue, since the ns2 server lives in the example.com zone.

> The Additional section would continue to serve glue records for the

Reviewer: This is intriguing but I don't quite believe it.  DELEG is in the parent zone, and is only retrieved by an explicit query (QTYPE=DELEG).  Sure, the parent should serve glue records alongside an NS record Answer, but why should it serve glue records alongside a DELEG Answer?

There's another wrinkle here though: SVCB recommends putting A+AAAA records for TargetName in the Additional section if they are in the same zone, to avoid a followup query.  If we apply that to DELEG, it could mean adding A+AAAA records for "ns2.example.com" (in the above example).  However, this would be redundant with the IP hints, and it would also be logically different: those A+AAAA records are locally available but they _aren't_ part of this zone.

Bottom line: we should do IP glue or IP hints, but not both.  Overall, my preference would be to stick with hints and get rid of the whole concept of glue.

Author: > is only retrieved by an explicit query (QTYPE=DELEG)

That's not really a settled matter.  I still think, pending testing to confirm that for the vast majority of the resolver base, that the real deployment win here is to serve NS, DS and DELEG together for any delegation and NOT require an explicit query for them.  We're trying to make the whole business of finding modern auths more efficient, not add more queries. 

And I really, really don't think we should be adding addresses for the DELEG names to the Additional section, because something that understands what DELEG is at all should be using the hints.

Reviewer: It seems like this is not settled, and is also not relevant to the transport draft.  Please remove it or move it to the other draft.

BTW, using QTYPE=DELEG doesn't add queries.  It just replaces the "arbitrary" QTYPE in QNAME minimization with a meaningful one.  Non-DELEG servers _should_ treat it as an unknown QTYPE and return the NS (and glue) in a delegation response.

resolvers.  Note that these could potentially be the same addresses as
are used in the hints of a DELEG record; NS and DELEG are orthogonal.

    ns1.example.com. 86400   IN  A      192.0.2.1
    ns1.example.com  86400   IN  AAAA   2001:DB8::1

As with SVCB, DELEG also offer the ability to use AliasMode for a
delegation. The example below shows that example.com is
being delegated with a DELEG AliasMode record which can then be
further resolved using SVCB to locate the actual parameters.

    example.com.  86400  IN DELEG 0 config2.example.net.
    example.com.  86400  IN NS      ns1.example.net.

The example.net authoritative server's SVCB record could look like this:

    config2.example.net 3600    IN SVCB ns2.example.net. (
                        alpn=dot
                        tlsa="2 0 1 ABC..."
                        ipv4hint=192.0.2.54
                        ipv6hint=2001:db8:2423::3 )

This document defines in detail the resolution process using these records.

## SvcParams

All SvcParamKeys for the "dns" scheme from SVCB-DNS {{?RFC9461}} apply as specified.  These are the "transport parameters", attributes describing how to reach an endpoint.

> The "alpn" transport parameter is OPTIONAL to include, unlike in SVCB-DNS, where it is generally required.  If the "alpn" SvcParamKey is omitted, the only available transport is presumed to be unencrypted DNS over UDP/TCP port 53.  Endpoints can indicate that insecure transport is not available by specifying "mandatory=alpn". [Why do we not require do53 to be explicit?]

Reviewer: Are you arguing that Do53 should be opt-in instead of opt-out?  We can certainly structure it that way; the difference seems pretty superficial.  I appreciate the current approach as an application of Occam's Razor, but it's possible that "mandatory=alpn" is too obscure or is stretching the "mandatory" semantics too far.

Author: Yes, I am arguing that if you are using a DELEG record, it should be explicit when you want do53 with it.  I don't want really anything assumed.  

It doesn't have to be in this draft initially though, since I can argue my case on the list.

(If all you wanted was Legacy DNS then just NS records.)

Reviewer: ```suggestion
The "alpn" transport parameter is OPTIONAL to include, unlike in SVCB-DNS, where it is generally required.  If the "alpn" SvcParamKey is omitted, the only available transport is presumed to be unencrypted DNS over UDP/TCP port 53.  Endpoints can indicate that insecure transport is not available by specifying "mandatory=alpn". [Should we change this to make do53 require a more explicit opt-in?]
```

Resolvers that support TLS-based transports MUST adopt one of the following behaviors:

1. Use DANE for authentication, and treat any endpoints lacking DANE support as incompatible.
2. Use PKI for authentication, and treat any DANE-only endpoint as incompatible.  In this behavior, compatibility with `tlsa=disabled` endpoints is REQUIRED.

> 3. Support both DANE and PKI for authentication, preferring DANE if it is available for each endpoint.

Reviewer: Markdown auto-increments numeric lists, so the numbers here have no effect.

Author: I thought so but I was editing on a plane and had no formatter so couldn't confirm.  Just out of curiosity, if you have numbered them but done so (apparently) incorrectly, does it fix them?  eg, 1, 2, 2, 4 ...

Reviewer: Yep, it completely ignores the actual numbers in the source code.  (This makes me sad: I want to be able to start with 0!)

;; services.  All resolvers can access it, but only resolvers with
;; DANE support will use DoQ.
alpn=doq tlsa="..."
[... how does this signal that it also offers do53? Is the though that it is
  because only this draft specifies alpn and the core draft doesn't (currently)

>   have a method to explicitly signal do53?  Maybe it should... ]

Reviewer: This server supports Do53 because it is not explicitly disclaimed by `mandatory=alpn`, as in the last example here.

Author: Okay, right, see above.  I think we should require signalling do53 to be explicit.

Reviewer: ```suggestion
[ If we move to requiring an explicit opt-in for do53 then these examples will have to change. ]
```

~~~
{: title="tlsa SvcParam Examples for DELEG"}

## Resolution procedure

> As specified in {{?I-D.draft-dnsop-deleg}}, the ServiceMode DELEG records provide a weak signal from the operator about which transport mechanisms they would most prefer clients to use, but it is not mandatory that resolvers try the delegated servers in priority order.  It might not even be possible for a resolver to do, such as if the highest priority record specifies an alpn that the resolver does not implement, but a lower priority record has an alpn that it can use.

Reviewer 2: I am still not sure if we actually want priority.

Author: I'm ambivalent on it.   While we still would need a way to distinguish AliasMode from ServiceMode, I'm kind of neutral about using a special magic 0 priority to do so.  I can see some limited use for it, but I also think it might be an attractive nuisance.  "Oh," says the Black hat,  "so the put this service at a lower priority?  Let's make that our  target!"

> As specified in {{?I-D.draft-dnsop-deleg}}, the ServiceMode DELEG records provide a weak signal from the operator about which transport mechanisms they would most prefer clients to use, but it is not mandatory that resolvers try the delegated servers in priority order.  It might not even be possible for a resolver to do, such as if the highest priority record specifies an alpn that the resolver does not implement, but a lower priority record has an alpn that it can use.

Reviewer: I don't understand this line.  Each ServiceMode record documents the supported transports for a particular endpoint, and the SvcPriority documents the recommended trial order.  Incompatible endpoints are skipped.  Of course clients are free to reorder them based on additional heuristics.  This is all covered in RFC 9460.

Author: What don't you understand then?  It's just making it clear in this document as well.   Do you think this statement is inconsistent with that, or should just be shorter in drawing attention to it?

Reviewer: OK.  If this is just meant to be highlighting some relevant aspects of RFC 9460, then it should say so.

Author: Will do.  I'll be working on some more editing tomorrow and will add the explicit reference.

Thus, as with authority server selection for NS records, a resolver is free to use its own policy implementation to determine which servers to continue to use for resolution.

If a connection error occurs when a resolver attempts to access a nameserver delegated to by a DELEG or SVCB record, such as with a certificate mismatch or by being unreachable, the resolver SHOULD attempt to connect to the other nameservers in the DELEG/SVCB RRset to until either exhausting the list or the resolver's policy indicates that they should treat the resolution as failed.

> For resolvers that strongly favor privacy, operators may wish to return a SERVFAIL when all attempts to use privacy-enhanced alpns in the DELEG/SVCB resolution process have failed. More permissive resolvers can fall back to using the legacy DNS resolution process via NS records.

Reviewer: I think this needs adjustment.  In this model, if I publish 5 DELEG records, and only one of them offers an encrypted endpoint, that single endpoint has to uphold my entire SLO.  (This could easily happen when delegating to multiple operators, or during a gradual rollout.)
```suggestion
Operators MAY return SERVFAIL when all attempts to use the DELEG/SVCB resolution process have failed.  This might have privacy benefits, especially if the DELEG records only offer encrypted transports.  More permissive resolvers can fall back to using the legacy DNS resolution process via NS records.
```

## Deployment Considerations

DELEG and SVCB records are intended to be the best method for modern resolvers to use delegation information for zones.

> Realistically, however, legacy resolvers will continue to exist on the Internet for a long time to come.  If a zone operator removes all NS records, the availability of their zones will be impacted for the clients that are using non-supporting resolvers. In some cases, this may be a desired quality, but should be noted by zone owners and operators.

Reviewer: Why is this in the transport draft?  This topic seems like it should go in the main DELEG draft.

Author: Agreed.  It was here because it was the document I was working on first on the airplane, and I was addressing "fill this section out" with relevant thoughts.

Reviewer: OK, please delete this section.

# Privacy Considerations

> All of the information transmitted by this protocol is public information published in the DNS.

Reviewer: I don't think queries are meant to be "published" in the DNS.

Author: I also thought the sentence as a little awkward, but was working with some prior text.   I agree regarding queries.   This sentence could maybe be improved with "transmitted by authoritative servers", but the section needs work anyway.

Reviewer: ```suggestion
Negotiation of alternative transports with DELEG allows the use of authenticated encryption between resolvers and authoritative DNS servers.  This improves privacy by concealing the QNAME and QTYPE from a network observer.  This is especially beneficial when a resolver is only shared by a small number of users.
```

# Security Considerations

> No new security risks are introduced by this specification.  [Too bold to say?]

Reviewer: Sure, maybe we aren't making anything worse (although I worry about DNS Water Torture-like attacks), but security and privacy considerations still need to cover how the protocol can fail to achieve the intended improvements.  In this case, we need to talk about how TLS is authenticated, especially in unsigned zones.  (Related: #5)

Author: Sounds good.  I wasn't really pushing for that sentence so much as putting a starting point that would draw out the "you were wrong on the Internet, and let me tell you why ..." brainstorming.
